File: backend/app/schemas/individual.py
import re
import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, EmailStr, field_validator


class VisaEntry(BaseModel):
    country: str
    visa_type: str  # "Tourist" | "Residence" | "Work" | "Student" | "Transit" | "Other"
    visa_number: str | None = None
    issue_date: date | None = None
    expiry_date: date | None = None


class IndividualCreate(BaseModel):
    full_name: str
    date_of_birth: date | None = None
    relationship_to_owner: str | None = None
    phone_number: str | None = None
    email: EmailStr | None = None
    address: str | None = None
    aadhaar_number: str | None = None
    pan_number: str | None = None
    passport_number: str | None = None
    passport_expiry: date | None = None
    driving_license_number: str | None = None
    voter_id_number: str | None = None
    skywards_number: str | None = None
    maharaja_number: str | None = None
    indigo_chip_number: str | None = None
    visas: list[VisaEntry] = []
    notes: str | None = None

    @field_validator("aadhaar_number")
    @classmethod
    def validate_aadhaar(cls, v: str | None) -> str | None:
        if v and (not v.isdigit() or len(v) != 12):
            raise ValueError("Aadhaar number must be exactly 12 digits")
        return v

    @field_validator("pan_number")
    @classmethod
    def validate_pan(cls, v: str | None) -> str | None:
        if v and not re.match(r"^[A-Z]{5}[0-9]{4}[A-Z]$", v.upper()):
            raise ValueError("PAN must be in format ABCDE1234F")
        return v.upper() if v else v


class IndividualUpdate(BaseModel):
    full_name: str | None = None
    date_of_birth: date | None = None
    relationship_to_owner: str | None = None
    phone_number: str | None = None
    email: EmailStr | None = None
    address: str | None = None
    photo_key: str | None = None
    aadhaar_number: str | None = None
    aadhaar_photo_key: str | None = None
    pan_number: str | None = None
    pan_photo_key: str | None = None
    passport_number: str | None = None
    passport_expiry: date | None = None
    passport_photo_key: str | None = None
    driving_license_number: str | None = None
    voter_id_number: str | None = None
    skywards_number: str | None = None
    maharaja_number: str | None = None
    indigo_chip_number: str | None = None
    visas: list[VisaEntry] | None = None
    notes: str | None = None

    @field_validator("aadhaar_number")
    @classmethod
    def validate_aadhaar(cls, v: str | None) -> str | None:
        if v and (not v.isdigit() or len(v) != 12):
            raise ValueError("Aadhaar number must be exactly 12 digits")
        return v

    @field_validator("pan_number")
    @classmethod
    def validate_pan(cls, v: str | None) -> str | None:
        if v and not re.match(r"^[A-Z]{5}[0-9]{4}[A-Z]$", v.upper()):
            raise ValueError("PAN must be in format ABCDE1234F")
        return v.upper() if v else v


class IndividualResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: uuid.UUID
    full_name: str
    date_of_birth: date | None = None
    relationship_to_owner: str | None = None
    phone_number: str | None = None
    email: str | None = None
    address: str | None = None
    photo_key: str | None = None
    photo_url: str | None = None
    aadhaar_number: str | None = None
    aadhaar_photo_url: str | None = None
    pan_number: str | None = None
    pan_photo_url: str | None = None
    passport_number: str | None = None
    passport_expiry: date | None = None
    passport_photo_url: str | None = None
    driving_license_number: str | None = None
    voter_id_number: str | None = None
    skywards_number: str | None = None
    maharaja_number: str | None = None
    indigo_chip_number: str | None = None
    visas: list[dict] = []
    notes: str | None = None
    is_archived: bool = False
    created_at: datetime | None = None

    # Computed by the service, not stored on the row.
    asset_count: int = 0
    active_visa_count: int = 0
    has_expiring_docs: bool = False

File: frontend/src/types/individual.ts
export interface VisaEntry {
  country: string;
  visa_type: string;
  visa_number?: string;
  issue_date?: string;
  expiry_date?: string;
}

export interface Individual {
  id: string;
  user_id: string;
  full_name: string;
  date_of_birth?: string;
  relationship_to_owner?: string;
  phone_number?: string;
  email?: string;
  address?: string;
  photo_url?: string;
  aadhaar_number?: string;
  aadhaar_photo_url?: string;
  pan_number?: string;
  pan_photo_url?: string;
  passport_number?: string;
  passport_expiry?: string;
  passport_photo_url?: string;
  driving_license_number?: string;
  voter_id_number?: string;
  skywards_number?: string;
  maharaja_number?: string;
  indigo_chip_number?: string;
  visas: VisaEntry[];
  notes?: string;
  is_archived: boolean;
  created_at?: string;
  asset_count: number;
  active_visa_count: number;
  has_expiring_docs: boolean;
}

export type IndividualCreate = {
  full_name: string;
  date_of_birth?: string;
  relationship_to_owner?: string;
  phone_number?: string;
  email?: string;
  address?: string;
  aadhaar_number?: string;
  pan_number?: string;
  passport_number?: string;
  passport_expiry?: string;
  driving_license_number?: string;
  voter_id_number?: string;
  skywards_number?: string;
  maharaja_number?: string;
  indigo_chip_number?: string;
  visas?: VisaEntry[];
  notes?: string;
};

export type IndividualUpdate = Partial<IndividualCreate> & {
  photo_key?: string;
  aadhaar_photo_key?: string;
  pan_photo_key?: string;
  passport_photo_key?: string;
};

/** doc_type accepted by POST /individuals/:id/documents/upload-url */
export type IdentityDocType = 'aadhaar' | 'pan' | 'passport' | 'photo';
